// mvdom/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.version = '0.8.6';
var event_1 = require("./event");
exports.on = event_1.on;
exports.off = event_1.off;
exports.trigger = event_1.trigger;
exports.bindDOMEvents = event_1.bindDOMEvents;
var dx_1 = require("./dx");
exports.push = dx_1.push;
exports.pull = dx_1.pull;
exports.pusher = dx_1.pusher;
exports.puller = dx_1.puller;
var dom_1 = require("./dom");
exports.first = dom_1.first;
exports.all = dom_1.all;
exports.closest = dom_1.closest;
exports.next = dom_1.next;
exports.prev = dom_1.prev;
exports.append = dom_1.append;
exports.frag = dom_1.frag;
var view_1 = require("./view");
exports.hook = view_1.hook;
exports.display = view_1.display;
exports.remove = view_1.remove;
exports.empty = view_1.empty;
var hub_1 = require("./hub");
exports.hub = hub_1.hub;
exports.bindHubEvents = hub_1.bindHubEvents;
exports.unbindHubEvents = hub_1.unbindHubEvents;
var utils_1 = require("./utils");
exports.val = utils_1.val;
require("./view-event.js");
require("./view-hub.js");
